src/test/java/spoon/test/template/SubstitutionTest.java

```diff
@@ -47,4 +51,44 @@ public void testSubstitutionInsertAllFields() {
         @Override
         public void statement() { }
     }
+
+    @Test
+    public void testSubstitutionInsertAllNestedTypes() {
+        // contract: Substitution.insertAllNestedTypes inserts the only nested class from a singly nested template into the target class
+
+        // arrange
+        Launcher spoon = new Launcher();
+        spoon.addTemplateResource(new FileSystemFile("./src/test/java/spoon/test/template/SubstitutionTest.java"));
+
+        spoon.buildModel();
+        Factory factory = spoon.getFactory();
+
+        CtType<?> targetType = factory.Class().create("goodClass");
+        StatementTemplate template = new SinglyNestedTemplate();
+
+        // act
+        Substitution.insertAllNestedTypes(targetType, template);
+
+        // assert
+        assertEquals(1, targetType.getNestedTypes().size());
+    }
+
+    private static class SinglyNestedTemplate extends StatementTemplate {
+
+        public class nestedClass<T extends Annotation> extends CtTypeImpl<T> {
+
+            @Override
+            public boolean isSubtypeOf(CtTypeReference<?> type) {
+                return false;
+            }
+
+            @Override
+            public void accept(CtVisitor visitor) {
+            }
+        }
```

Comment:
As with your first attempt with the last template, there's a lot going on here that doesn't seem related to the test. To pass the test, the minimal required nested type is this:

```suggestion
        class nestedClass {
        }
```

In general, make tests as minimal as possible, and don't include things that aren't used. Is there any reason for including the other things? If there is, then you should attempt to use them in assertions such that it's obvious they're there for a reason.

Reply:
yeah actually originally I was aiming for stronger assertion than size, but wasn't able to think of it 😅